src/main/java/com/linkedin/kafka/clients/consumer/ConsumerRecordsProcessor.java

```diff
@@ -333,34 +314,60 @@ public void close() {
+      LOG.trace("Skipping message {} from partition {} because its offset is smaller than the high watermark",
+          srcRecord.offset(), topicPartition);
+      _deliveredMessageOffsetTracker.track(topicPartition, srcRecord.offset(), safeOffset, srcRecord.offset(), false);
+      return null;
     }
-    return handledRecord;
-  }
 
-  private byte[] parseAndMaybeTrackRecord(TopicPartition tp, long messageOffset, byte[] bytes) {
-    MessageAssembler.AssembleResult assembledResult = _messageAssembler.assemble(tp, messageOffset, bytes);
-    if (assembledResult.messageBytes() != INCOMPLETE_RESULT) {
-      LOG.trace("Got message {} from partition {}", messageOffset, tp);
-      boolean shouldSkip = shouldSkip(tp, messageOffset);
-      // The safe offset is the smaller one of the current message offset + 1 and current safe offset.
-      long safeOffset = Math.min(messageOffset + 1, _messageAssembler.safeOffset(tp));
-      _deliveredMessageOffsetTracker.track(tp,
-                                           messageOffset,
-                                           safeOffset,
-                                           assembledResult.messageStartingOffset(),
-                                           !shouldSkip);
-      // We skip the messages whose offset is smaller than the high watermark.
-      if (shouldSkip) {
-        LOG.trace("Skipping message {} from partition {} because its offset is smaller than the high watermark",
-                  messageOffset, tp);
-        return INCOMPLETE_RESULT;
-      } else {
-        return assembledResult.messageBytes();
-      }
-    } else {
-      _deliveredMessageOffsetTracker.addNonMessageOffset(tp, messageOffset);
-      return INCOMPLETE_RESULT;
+    if (assembledResult == null) {
+      //Not a large message segment
+      _deliveredMessageOffsetTracker.track(topicPartition, srcRecord.offset(), safeOffset, srcRecord.offset(), true);
+      return srcRecord;
+    }
+
+    if (assembledResult.messageBytes() == null) {
+      //Not a complete, large message
+      _deliveredMessageOffsetTracker.addNonMessageOffset(topicPartition, srcRecord.offset());
+      return null;
     }
+
+    // Completed, large message value
+    _deliveredMessageOffsetTracker.track(topicPartition, srcRecord.offset(), safeOffset, assembledResult.messageStartingOffset(),
+        true);
+
+    int serializedKeySize = assembledResult.isOriginalKeyNull() ? 0 : srcRecord.key().length;
+    byte[] key = assembledResult.isOriginalKeyNull() ? null : srcRecord.key();
+    int serializedValueSize = assembledResult.messageBytes().length;
+
+    ExtensibleConsumerRecord<byte[], byte[]> largeMessageRecord =
+      new ExtensibleConsumerRecord<>(srcRecord.topic(), srcRecord.partition(), srcRecord.offset(),
+        srcRecord.timestamp(), srcRecord.timestampType(),
+        srcRecord.checksum(),
+        serializedKeySize, serializedValueSize,
+        key, assembledResult.messageBytes());
+    //TODO: checksums recomputed?
```

Comment:
It's not clear to me why someone would want the checksum since they would need to recompute the original bytes that were sent in order to verify the checksum and presumably the underlying code has already done that.